clients/da-vinci-client/src/main/java/com/linkedin/davinci/helix/LeaderFollowerPartitionStateModel.java

```diff
@@ -195,12 +195,27 @@ public void onBecomeDroppedFromOffline(Message message, NotificationContext cont
           // Gracefully drop partition to drain the requests to this partition
           Thread.sleep(TimeUnit.SECONDS.toMillis(getStoreAndServerConfigs().getPartitionGracefulDropDelaySeconds()));
         } catch (InterruptedException e) {
-          throw new VeniceException("Got interrupted during state transition: 'OFFLINE' -> 'DROPPED'", e);
+          throw new VeniceException("Got interrupted while waiting for graceful drop delay of serving version", e);
         } finally {
           this.threadPoolStats.decrementThreadBlockedOnOfflineToDroppedTransitionCount();
         }
       }
-      removePartitionFromStoreGracefully();
+      CompletableFuture<Void> dropPartitionFuture = removePartitionFromStoreGracefully();
+      boolean waitForDropPartition = !dropPartitionFuture.isDone();
+      try {
+        if (waitForDropPartition) {
+          this.threadPoolStats.incrementThreadBlockedOnOfflineToDroppedTransitionCount();
+        }
+        dropPartitionFuture.get(WAIT_DROP_PARTITION_TIME_OUT_MS, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException e) {
+        throw new VeniceException("Got interrupted while waiting for drop partition future to complete", e);
```

Comment:
When time-out happened, we will throw exception here, but not in the ERROR->DROPPED state transition above. Do we want to make them consistent or they are totally different and we need separate ways of handling them?

Reply:
They are different, we explicitly don't want to throw exception from ERROR->DROPPED to prevent us from getting into an infinite loop when we have bad internal state. i.e. ERROR->DROPPED->ERROR->......

Reply:
SG. Can we log the error messages for debugging purpose, in case the timeout happens?

Reply:
Which function are you referring to for ERROR->DROPPED? I see `onBecomeOfflineFromError` but it's a No-op. @lluwm

Reply:
No, it's the `AbstractPartitionStateModel.onBecomeDroppedFromError` that's also modified in this PR.